### ISSUE TITLE
Cancelled (Dio) HTTP requests show as errors in RUM

### ISSUE THREAD

**Issue by scopendo** [DataDog/dd-sdk-flutter#175]
**Describe what happened**

My application uses the `Dio` library to make HTTP requests.  This allows for requests to be cancelled, which is useful when performing a typeahead search to cancel a previously dispatched API call when the search input changes.

It appears that when a Dio-dispatched request is cancelled, that `datadog_tracking_http_client` records an error in RUM, which adds unwanted noise.

**Steps to reproduce the issue:**

Here is an example using Riverpod, whereby a search for staff members is dispatched when search input changes, with the request automatically being cancelled when the provider is no longer watched, which happens when the search input changes.

```dart
  ...
  @override
  Widget build(BuildContext context, WidgetRef ref) {
    ...
    final searchProvider = staffMemberSearchFamily(searchText.value);
    final matches = ref.watch(searchPod);
    ...
  }
  ...
```

```dart
final staffMemberSearchFamily = FutureProvider.family.autoDispose<List<PersonData>, String?>(
  (ref, searchText) async {
    // Dispatch a cancellable HTTP request, meaning that we instruct the HTTP
    // client to ignore the response.
    final cancelToken = CancelToken();
    ref.onDispose(cancelToken.cancel);

    return ref.watch(personApiPod).internalPersonSearch(
          searchText: searchText,
          cancelToken: cancelToken,
        );
  },
);
```

**Describe what you expected:**

I expected that programmatically-cancelled HTTP requests would not appear in RUM as errors.

**Additional context**

 - Dart/Flutter version: 2.17.6/3.0.5
 - Datadog SDK version: 1.0.0-beta.3
 - datadog_tracking_http_client version: 1.0.1-beta.1
 - Versions of any other relevant dependencies: Dio 4.0.6

**Comment by fuzzybinary** [comment]
Hi @scopendo! 

Thanks for reporting this. I'll start looking into this. It looks like Dio calls `HttpClient.close(force: true)`, when you cancel, which likely throws an error on the Request or Response object that we'll have to handle.

Are the errors always the same? Could you send me over what errors you're seeing?

**Comment by scopendo** [comment]
Sorry, I meant to add that before.  The vast majority of errors are the same:

> HttpException: HttpException: Connection closed before full header was received, uri = https://...

From the `error` attributes:

```
isCrash: false
type: HttpException
```

**Comment by fuzzybinary** [comment]
Thanks @scopendo!

I was hoping it was something less generic that I could easily filter out as a non-error. But it looks like what we'll have to do is keep track of whether the client was closed and ignore any errors after closure, as that exception could be a legitimate error we'd want to track.

I'll look into the best approach.

**Comment by fuzzybinary** [comment]
Hi @scopendo

Sorry for the delay on this, but for the moment, Datadog doesn't have a good way to mark a request as "canceled." We're still looking into the best approach on the Native side, which we'll then bring into the Flutter side as soon as its available.

**Comment by scopendo** [comment]
@fuzzybinary – I know it's a bit crappy but would it be possible to suppress any exceptions containing the text "Connection closed before full header was received"?  Or is there a way for me to filter the error myself, as I believe is possible in the browse SDK?

**Comment by fuzzybinary** [comment]
I'm worried about filtering that exception only because it could have other causes. For that matter, I *believe* you could get other errors during cancelation, depending on the timing of the cancelation.

Unfortunately, filtering is not possible on the Flutter side at the moment, but it is on the roadmap. I'm looking for ways to support filtering sooner rather than later and will keep you posted on our progress.

**Comment by scopendo** [comment]
Hi @fuzzybinary – is there any progress on the native SDK side for supporting this?

**Comment by fuzzybinary** [comment]
We are working on supporting mapping in the Flutter SDK very soon, which should allow you to inspect, modify, and discard events before they get sent to Datadog. This will likely be our recommend way for filtering out the cancelations.

Sorry it's taken so long and thanks for your patience!

**Comment by scopendo** [comment]
Hi @fuzzybinary – just checking in on this.

**Comment by fuzzybinary** [comment]
@scopendo I know I say this every time, but "soon."

We actually have log mapping in current source, and RUM should be following shortly behind. It's one of my top priorities in the new year.

Really sorry for the delay!

**Comment by fuzzybinary** [comment]
This is now in and shipped as part of 1.2.0.  The mapper you'll want is the `rumErrorEventMapper` which will contain the resource that it attempted to load and the error that resulted. Return null from the mapper and it will discard the event (it won't be sent to Datadog and won't appear in your RUM sessions).

Let me know if you have any issues!
